### PR TITLE
Generic metadata prespecialization: enums

### PR DESCRIPTION
**Pull request by nate-chandler**
Continuing the work started in https://github.com/apple/swift/pull/28610 , prespecializes generic metadata for enums under limited circumstances. Specifically, prespecialization only occurs for usages within the same module where the type itself is defined and for specializations all of whose type arguments are themselves concrete types.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

File: lib/IRGen/GenMeta.cpp [hunk 1]
```diff
@@ -3789,10 +3806,34 @@ namespace {
 
       return flags;
     }
+  };
+
+  // FIXME: Without this template typealias, the following errors are produced
```
Comment (slavapestov) [inline lib/IRGen/GenMeta.cpp]: Interesting, is there a radar for this?

File: lib/IRGen/MetadataRequest.cpp [hunk 2]
```diff
@@ -684,11 +684,6 @@ bool irgen::isNominalGenericContextTypeMetadataAccessTrivial(
     return false;
```
Comment (slavapestov) [inline lib/IRGen/MetadataRequest.cpp]: I think this is too conservative. nominal.isResilient(..., ResilienceExpansion::Minimal) ignores the module argument and returns true if the module is resilient.

But if its resilient and defined in the same module, you can still pre-specialize it. The resilience means the metadata is opaque across module boundaries, not within the module itself.

**Comment by slavapestov** [comment]
I can't comment on unchanged lines in the diff, but about this part:

```
    // For now, to avoid statically specializing generic protocol witness
    // tables, don't statically specialize metadata for types any of whose
    // arguments are generic.
    //
    // TODO: This is more pessimistic than necessary.  Specialize even in
    //       the face of generic arguments so long as those arguments
    //       aren't required to conform to any protocols.
    //
    // TODO: Once witness tables are statically specialized, check whether the
    //       ConformanceInfo returns nullptr from tryGetConstantTable.
    //       early return.
```

Can you change this to use the same predicates as the code for emitting witness table references, where it decides if it should reference a witness table directly or call an accessor function? I think `isDependentConformance()` is the right check, or at least a starting point.

File: test/IRGen/prespecialized-metadata/enum-inmodule-1argument-1distinct_use.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// CHECK: @"$s4main5ValueOySiGMf" = internal constant <{ i8**, [[INT]], %swift.type_descriptor*, %swift.type*, i64 }> <{ i8** getelementptr inbounds (%swift.enum_vwtable, %swift.enum_vwtable* @"$s4main5ValueOySiGWV", i32 0, i32 0), [[INT]] 513, %swift.type_descriptor* bitcast (<{ i32, i32, i32, i32, i32, i32, i32, i32, i32, i16, i16, i16, i16, i8, i8, i8, i8 }>* @"$s4main5ValueOMn" to %swift.type_descriptor*), %swift.type* @"$sSiN", i64 3 }>, align [[ALIGNMENT]]
+enum Value<First> {
```
Comment (slavapestov) [inline test/IRGen/prespecialized-metadata/enum-inmodule-1argument-1distinct_use.swift]: Can you add some tests that use protocol conformances, both the supported and unsupported cases?

Comment (nate-chandler) [reply test/IRGen/prespecialized-metadata/enum-inmodule-1argument-1distinct_use.swift]: Yes, I'm in the process of writing more.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test platform linux

**Comment by nate-chandler** [comment]
@swift-ci please test platform windows

**Comment by nate-chandler** [comment]
@swift-ci please test linux platform

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

**Comment by nate-chandler** [comment]
@swift-ci please clean test linux platform

**Comment by nate-chandler** [comment]
@swift-ci please clean test windows platform

**Comment by nate-chandler** [comment]
@swift-ci please clean test linux platform

**Comment by nate-chandler** [comment]
@swift-ci please clean test linux platform

File: lib/IRGen/GenMeta.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/IRGen/GenMeta.cpp]: Unfortunately, I wasn't able to minimally recreate the problem, but rdar://problem/58884416 covers this.

**Comment by nate-chandler** [comment]
> Can you change this to use the same predicates as the code for emitting witness table references, where it decides if it should reference a witness table directly or call an accessor function? I think isDependentConformance() is the right check, or at least a starting point.

@slavapestov, thanks for pointing me at that!  Is this along the lines of what you have in mind?  https://github.com/apple/swift/pull/28692/commits/b71623560e425015448fa4e8172b16f9127838b5

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test windows platform

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Pull request merged**